sql: crash when collecting stats on virtual computed column with enum type

In https://github.com/cockroachdb/cockroach/pull/122313 we fixed a missing type resolver when planning stats collection on virtual computed columns with enum types. But something is still not quite right, because now the following [repro](https://gist.github.com/codingconcepts/0b8f2ae5c1f31aa4a241d92e6f473841) crashes the node with a nil pointer dereference during type checking.

```sql
CREATE TYPE order_status AS ENUM ('pending', 'paid', 'dispatched', 'delivered');

CREATE TABLE orders (
  "id" UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  "customer_id" UUID NOT NULL,
  "total" DECIMAL NOT NULL,
  "balance" DECIMAL NOT NULL,
  "order_ts" TIMESTAMPTZ(0) NOT NULL DEFAULT now(),
  "dispatch_ts" TIMESTAMPTZ(0),
  "delivery_ts" TIMESTAMPTZ(0),
  "status" order_status AS (
    CASE
      WHEN "delivery_ts" IS NOT NULL THEN 'delivered'
      WHEN "dispatch_ts" IS NOT NULL THEN 'dispatched'
      WHEN "balance" = 0 THEN 'paid'
      ELSE 'pending'
    END) VIRTUAL,

  INDEX ("status")
) WITH (sql_stats_automatic_collection_enabled=false);

INSERT INTO orders ("customer_id", "total", "balance", "dispatch_ts", "delivery_ts") VALUES
  ('bdeb232e-12e9-4a33-9dd5-7bb9b694291a', 100, 100, NULL, NULL),
  ('0dc59725-d20b-4370-a05d-11db025a0064', 200, 0, NULL, NULL),
  ('d53d4021-9390-4b3a-9e5a-4bf1ff3e5a4c', 300, 0, now(), NULL),
  ('d53d4021-9390-4b3a-9e5a-4bf1ff3e5a4c', 400, 0, now(), now());

SET CLUSTER SETTING sql.stats.virtual_computed_columns.enabled = 'on';

ANALYZE orders;
```

Here's the stack trace on master (e0d587ccfccd7e95ff1e4b96d602356db2a247ea):

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x103750c28]

goroutine 4550 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x1400529bac8?, {0x10979e250, 0x14004a16d80})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x68
panic({0x108d762e0?, 0x10ce759e0?})
	GOROOT/src/runtime/panic.go:770 +0x124
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).ReadTimestamp(0x1400529bca8?)
	github.com/cockroachdb/cockroach/pkg/kv/txn.go:401 +0x28
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*leasedDescriptors).getByID(0x14005667208, {0x10979e250, 0x14005760690}, {0x10976b0c0, 0x0}, 0x68)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/leased_descriptors.go:121 +0x64
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*byIDLookupContext).lookupLeased(0x1400529be00, 0x68)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/descriptor.go:376 +0xc8
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.getDescriptorsByID({0x10979e250, 0x14005760690}, 0x140056671e0, 0x0, {{0x8?, 0x0?}, {0x0?, 0x0?, 0x0?, 0x0?}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/descriptor.go:180 +0x74c
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.ByIDGetter.Desc({{0x0?, 0x140056671e0?}, {{0x80?, 0x0?}, {0x0?, 0x0?, 0x0?, 0x0?}, {0x8?, 0x9?, ...}}}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/getters.go:48 +0xc8
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.ByIDGetter.Type({{0x0?, 0x140056671e0?}, {{0xd8?, 0xc0?}, {0x29?, 0x5?, 0x40?, 0x1?}, {0x18?, 0x38?, ...}}}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/getters.go:106 +0x38
github.com/cockroachdb/cockroach/pkg/sql.(*schemaResolver).GetTypeDescriptor(0x1400535c608, {0x10979e250, 0x14005760690}, 0x38f3e20?)
	github.com/cockroachdb/cockroach/pkg/sql/schema_resolver.go:379 +0xac
github.com/cockroachdb/cockroach/pkg/sql/catalog/typedesc.ResolveHydratedTByOID({0x10979e250, 0x14005760690}, 0x1668ebc8?, {0x109755aa0, 0x1400535c608})
	github.com/cockroachdb/cockroach/pkg/sql/catalog/typedesc/hydrate.go:39 +0x84
github.com/cockroachdb/cockroach/pkg/sql.(*schemaResolver).ResolveTypeByOID(...)
	github.com/cockroachdb/cockroach/pkg/sql/schema_resolver.go:371
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.ResolveType({0x10979e250?, 0x14005760690?}, {0x10974ece0?, 0x140075ae264?}, {0x10976ae40?, 0x1400535c608?})
	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/type_name.go:145 +0x1a8
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*AnnotateTypeExpr).TypeCheck(0x14005761080, {0x10979e250, 0x14005760690}, 0x1400535c9e8, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/type_check.go:770 +0x68
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.typeCheckSameTypedExprs({0x10979e250, 0x14005760690}, 0x1400535c9e8, 0x140039260c0, {0x14004c5dc00, 0x4, 0x4})
	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/type_check.go:2832 +0x4f0
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*CaseExpr).TypeCheck(0x14004c5dbc0, {0x10979e250, 0x14005760690}, 0x1400535c9e8, 0x140039260c0)
	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/type_check.go:503 +0x3cc
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.TypeCheck({0x10979e250?, 0x14005760690?}, {0x10979ecd0?, 0x14004c5dbc0?}, 0x7?, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/type_check.go:350 +0x98
github.com/cockroachdb/cockroach/pkg/sql/catalog/schemaexpr.MakeComputedExprs({0x10979e250, 0x14005760690}, {0x14002074c60, 0x1, 0x140028e5ea0?}, {0x14002c58fc0, 0x7, 0x7}, {0x10988e860, 0x14003a76c08}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/catalog/schemaexpr/computed_column.go:256 +0x47c
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createStatsPlan(_, {_, _}, _, _, {_, _}, {_, _, _}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_plan_stats.go:457 +0x4c4
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPlanForCreateStats(_, {_, _}, _, _, _, {{0x0, 0x0}, {{0x1400192d274, 0x6}, ...}, ...})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_plan_stats.go:634 +0x394
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).planAndRunCreateStats(0x14002380640, {0x10979e250?, 0x14004a170e0?}, 0x1400535cad8, 0x14003a32dd0, 0x1400535c9e8, 0x1400644ba40, 0x140058dd380, 0x140057e87c0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_plan_stats.go:649 +0x114
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsResumer).Resume.func1({0x10979e250, 0x14004a170e0}, {0x109806480, 0x1400644bb80})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:672 +0x264
github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn.func1({0x10979e250?, 0x14004a170e0?}, 0x1098083b8?)
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1794 +0x38
github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn.func4({0x10979e250, 0x14004a170e0}, 0x1400644ba40)
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1881 +0x260
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0x1400644ba40, {0x10979e250, 0x14004a170e0}, 0x14004c5d700)
	github.com/cockroachdb/cockroach/pkg/kv/txn.go:1049 +0x74
github.com/cockroachdb/cockroach/pkg/kv.runTxn({0x10979e250, 0x14004a170e0}, 0x1400644ba40, 0x1?)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:1089 +0x3c
github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl(0x1400529ee28?, {0x10979e250, 0x14004a170e0}, 0x9401380?, 0x1?, 0x0, 0x14004c5d700)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:1052 +0xa8
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(...)
	github.com/cockroachdb/cockroach/pkg/kv/db.go:1027
github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn(0x140045b0f30, {0x10979e250, 0x14004a170e0}, 0x140020749b0, {0x0, 0x0, 0x140058dd301?})
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1868 +0x230
github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn(0x140045b0f30, {0x10979e250, 0x14004a170e0}, 0x14004c5d6c0, {0x0, 0x0, 0x0})
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:1795 +0x90
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsResumer).Resume(0x1400651efb0, {0x10979e250, 0x14004a170e0}, {0x1093edb80?, 0x1400354d910})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:655 +0x2a0
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2(0x140058dd380?, 0x0?, 0x1400529f8f8, {0x10978ca90?, 0x1400651efb0?}, {0x10979e250?, 0x14004a170e0?}, {0x1093edb80?, 0x1400354d910?})
	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1632 +0xec
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0x14003a80008, {0x10979e250, 0x14004a17020}, {0x1093edb80, 0x1400354d910}, {0x10978ca90, 0x1400651efb0}, 0x140058dd380, {0x106f867ee, 0x7}, ...)
	github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1633 +0x83c
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0x14003a80008, {0x10979e250, 0x14004a17020}, {0x10978ca90, 0x1400651efb0}, 0x140058dd380, {0x106f867ee, 0x7}, {0x1400858c888, 0x16})
	github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:456 +0x3e4
github.com/cockroachdb/cockroach/pkg/jobs.(*StartableJob).Start.func2({0x1400416d808?, 0x14004a16750?})
	github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:828 +0xdc
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:485 +0x128
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 3198
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:476 +0x31c
# michae2@michae2-crl-m3 2 ~crdb %
```

Jira issue: CRDB-38560

--- a/pkg/sql/create_stats.go
+++ b/pkg/sql/create_stats.go
@@ -99,6 +99,12 @@ func StubTableStats(
 // framework.
 type createStatsNode struct {
 	tree.CreateStats
+
+	// p is the "outer planner" from planning the CREATE STATISTICS
+	// statement. When we startExec the createStatsNode, it creates a job which
+	// has a second planner (the JobExecContext). When the job resumes, it does
+	// its work using a retrying internal transaction for which we create a third
+	// "inner planner".
 	p *planner
 
 	// runAsJob is true by default, and causes the code below to be executed,
@@ -638,39 +644,53 @@ var _ jobs.Resumer = &createStatsResumer{}
 
 // Resume is part of the jobs.Resumer interface.
 func (r *createStatsResumer) Resume(ctx context.Context, execCtx interface{}) error {
-	p := execCtx.(JobExecContext)
+	// jobsPlanner is a second planner distinct from the "outer planner" in the
+	// createStatsNode. It comes from the jobs system and does not have an
+	// associated txn.
+	jobsPlanner := execCtx.(JobExecContext)
 	details := r.job.Details().(jobspb.CreateStatsDetails)
 	if details.Name == jobspb.AutoStatsName {
 		// We want to make sure that an automatic CREATE STATISTICS job only runs if
 		// there are no other CREATE STATISTICS jobs running, automatic or manual.
-		if err := checkRunningJobs(ctx, r.job, p); err != nil {
+		if err := checkRunningJobs(ctx, r.job, jobsPlanner); err != nil {
 			return err
 		}
 	}
 
 	r.tableID = details.Table.ID
-	evalCtx := p.ExtendedEvalContext()
+	evalCtx := jobsPlanner.ExtendedEvalContext()
 
-	dsp := p.DistSQLPlanner()
-	if err := p.ExecCfg().InternalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
-		// Set the transaction on the EvalContext to this txn. This allows for
-		// use of the txn during processor setup during the execution of the flow.
-		evalCtx.Txn = txn.KV()
-
+	if err := jobsPlanner.ExecCfg().InternalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		// We create a third "inner planner" associated with this txn in order to
+		// have (a) use of the txn during type checking of any virtual computed
+		// column expressions, and (b) use of the txn during processor setup during
+		// the execution of the flow.
+		innerPlanner, cleanup := NewInternalPlanner(
+			"create-stats-resume-job",
+			txn.KV(),
+			jobsPlanner.User(),
+			&MemoryMetrics{},
+			jobsPlanner.ExecCfg(),
+			jobsPlanner.SessionData(),
+		)
+		defer cleanup()
+		innerP := innerPlanner.(*planner)
+		innerEvalCtx := innerP.ExtendedEvalContext()
 		if details.AsOf != nil {
-			p.ExtendedEvalContext().AsOfSystemTime = &eval.AsOfSystemTime{Timestamp: *details.AsOf}
-			p.ExtendedEvalContext().SetTxnTimestamp(details.AsOf.GoTime())
+			innerP.ExtendedEvalContext().AsOfSystemTime = &eval.AsOfSystemTime{Timestamp: *details.AsOf}
+			innerP.ExtendedEvalContext().SetTxnTimestamp(details.AsOf.GoTime())
 			if err := txn.KV().SetFixedTimestamp(ctx, *details.AsOf); err != nil {
 				return err
 			}
 		}
 
-		planCtx := dsp.NewPlanningCtx(ctx, evalCtx, nil /* planner */, txn.KV(), FullDistribution)
+		dsp := innerP.DistSQLPlanner()
+		planCtx := dsp.NewPlanningCtx(ctx, innerEvalCtx, innerP, txn.KV(), FullDistribution)
 		// CREATE STATS flow doesn't produce any rows and only emits the
 		// metadata, so we can use a nil rowContainerHelper.
 		resultWriter := NewRowResultWriter(nil /* rowContainer */)
 		if err := dsp.planAndRunCreateStats(
-			ctx, evalCtx, planCtx, p.SemaCtx(), txn.KV(), r.job, resultWriter,
+			ctx, innerEvalCtx, planCtx, innerP.SemaCtx(), txn.KV(), r.job, resultWriter,
 		); err != nil {
 			// Check if this was a context canceled error and restart if it was.
 			if grpcutil.IsContextCanceled(err) {

--- a/pkg/sql/logictest/testdata/logic_test/distsql_stats
+++ b/pkg/sql/logictest/testdata/logic_test/distsql_stats
@@ -2877,3 +2877,72 @@ vectorized: true
           estimated row count: 10 (1.0% of the table; stats collected <hidden> ago)
           table: mno@mno_o_idx
           spans: [/11 - /11]
+
+# Regression for not setting the TypeResolver on the SemaContext when dealing
+# with stats on virtual computed columns (#122312).
+statement ok
+CREATE TABLE t122312 (s STRING, g greeting AS (s::greeting) VIRTUAL)
+
+statement ok
+ANALYZE t122312
+
+# Regression for not setting the txn of the schemaResolver when type checking
+# stats on virtual computed columns.
+statement ok
+CREATE TYPE order_status AS ENUM ('pending', 'paid', 'dispatched', 'delivered')
+
+statement ok
+CREATE TABLE orders (
+  "id" UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+  "customer_id" UUID NOT NULL,
+  "total" DECIMAL NOT NULL,
+  "balance" DECIMAL NOT NULL,
+  "order_ts" TIMESTAMPTZ(0) NOT NULL DEFAULT now(),
+  "dispatch_ts" TIMESTAMPTZ(0),
+  "delivery_ts" TIMESTAMPTZ(0),
+  "status" order_status AS (
+    CASE
+      WHEN "delivery_ts" IS NOT NULL THEN 'delivered'
+      WHEN "dispatch_ts" IS NOT NULL THEN 'dispatched'
+      WHEN "balance" = 0 THEN 'paid'
+      ELSE 'pending'
+    END) VIRTUAL,
+  INDEX ("status")
+)
+
+statement ok
+INSERT INTO orders ("customer_id", "total", "balance", "dispatch_ts", "delivery_ts") VALUES
+  ('bdeb232e-12e9-4a33-9dd5-7bb9b694291a', 100, 100, NULL, NULL),
+  ('0dc59725-d20b-4370-a05d-11db025a0064', 200, 0, NULL, NULL),
+  ('d53d4021-9390-4b3a-9e5a-4bf1ff3e5a4c', 300, 0, now(), NULL),
+  ('d53d4021-9390-4b3a-9e5a-4bf1ff3e5a4c', 400, 0, now(), now())
+
+statement ok
+ANALYZE orders
+
+query TIIIIB colnames
+SELECT column_names, row_count, null_count, distinct_count, avg_size, histogram_id IS NOT NULL AS has_histogram
+FROM [SHOW STATISTICS FOR TABLE orders]
+ORDER BY column_names::STRING
+----
+column_names   row_count  null_count  distinct_count  avg_size  has_histogram
+{balance}      4          0           2               32        true
+{customer_id}  4          0           3               16        true
+{delivery_ts}  4          3           2               6         true
+{dispatch_ts}  4          2           2               12        true
+{id}           4          0           4               16        true
+{order_ts}     4          0           1               24        true
+{status}       4          0           4               48        true
+{total}        4          0           4               32        true
+
+let $hist_status
+SELECT histogram_id FROM [SHOW STATISTICS FOR TABLE orders] WHERE column_names = ARRAY['status']
+
+query TIRI colnames,nosort
+SHOW HISTOGRAM $hist_status
+----
+upper_bound   range_rows  distinct_range_rows  equal_rows
+'pending'     0           0                    1
+'paid'        0           0                    1
+'dispatched'  0           0                    1
+'delivered'   0           0                    1

